### PR TITLE
Use with_reporter() internally

### PR DESCRIPTION
**Pull request by krlmlr**
for better control over the testing logic (e.g., stop if encountering any error: #308).

In the long run, I think the code that's now in test_file() should be vectorized to allow for several files (perhaps extract the logic to a new function test_file_one()). Then test_files() simply calls test_file().

### CODE REVIEW

File: R/test-files.r [hunk 1]
```diff
@@ -40,10 +40,11 @@ test_files <- function(paths, reporter = "summary",
   if (length(paths) == 0) stop('No matching test file in dir')
 
   current_reporter <- find_reporter(reporter)
-  current_reporter$start_reporter()
-  results <- lapply(paths, test_file, env = env,
-    reporter = current_reporter, start_end_reporter = FALSE)
-  current_reporter$end_reporter()
+  with_reporter(
```
Comment (hadley) [inline R/test-files.r]: I think this would be clearer as `results <- with_reporter(current_reporter, ...)`

**Comment by hadley** [comment]
Otherwise, looks good.

File: R/test-files.r (hunk 1 above)
Comment (krlmlr) [inline R/test-files.r]: with_reporter() returns the reporter, not the results of the code. At least one package (checkmate) depends on this behavior. I'll add documentation.

**Comment by krlmlr** [comment]
Documentation added.

**Comment by hadley** [comment]
LGTM. Can you please add a news bullet and merge?

**Pull request merged**